### PR TITLE
A better alternative to #46344

### PR DESCRIPTION
**Pull request by rschu1ze**
The (experimental) inverted index writes/reads files different from the standard files written by the other skip indexes. The original problem was that with database engine "ordinary", DROP TABLE of a table with inverted index finds unknown files in persistence and complains. The same will happen with engine "atomic" but deferred. As a hotfix, the error was silenced in #46344 by explicitly adding the four files created in a specific test to the deletion code.

This PR tries a cleaner solution where all needed files are provided via the normal checksum mechanism. One drawback remains which is that the affected files were written earlier and we don't have their checksums available. Therefore, the inverted index is currently excluded from CHECK TABLE (but it was like that also before).

Minimal repro:

```sql
  SET allow_experimental_inverted_index = 1;
  DROP TABLE IF EXISTS tab;
  CREATE TABLE tab(s String, INDEX af(s) TYPE inverted(2)) ENGINE = MergeTree() ORDER BY s;
  INSERT INTO tab VALUES ('Alick a01');
  CHECK TABLE tab;
  DROP TABLE IF EXISTS tab;
```
Then run `./clickhouse-test` with `--db-engine Ordinary`

### Changelog category (leave one):
- Not for changelog

### CODE REVIEW

**Comment by UnamedRus** [comment]
> Therefore, the inverted index is currently excluded from CHECK TABLE (but it was like that also before).

AFAIK, ClickHouse recalculate checksums if some files does not exist in checksum.txt.

So it's completely OK?

**Comment by rschu1ze** [comment]
> AFAIK, ClickHouse recalculate checksums if some files does not exist in checksum.txt.

True.

> So it's completely OK?

Well, "Storages/MergeTree/MergeTreeDataPartChecksum.cpp" was changed to suppress that. My local testing was green. The whole situation isn't great right now but better than before and hopefully only an interim solution.

File: src/Storages/MergeTree/MergeTreeDataPartWriterOnDisk.cpp [hunk 1]
```diff
@@ -388,6 +388,18 @@ void MergeTreeDataPartWriterOnDisk::fillSkipIndicesChecksums(MergeTreeData::Data
         auto & stream = *skip_indices_streams[i];
         if (!skip_indices_aggregators[i]->empty())
             skip_indices_aggregators[i]->getGranuleAndReset()->serializeBinary(stream.compressed_hashing);
+
+        /// Register additional files written only by the inverted index. Required because otherwise DROP TABLE complains about unknown
+        /// files. Note that the provided actual checksums are bogus. The problem is that at this point the file writes happened already and
+        /// we'd need to re-open + hash the files (fixing this is TODO). For now, CHECK TABLE skips these four files.
+        if (dynamic_cast<const MergeTreeIndexInverted *>(&*skip_indices[i]) != nullptr)
```
Comment (alexey-milovidov) [inline src/Storages/MergeTree/MergeTreeDataPartWriterOnDisk.cpp:395]: dynamic_cast -> typeid_cast

**Review by tavplubix — APPROVED** [review]

**Comment by tavplubix** [comment]
It would be great to add a test with a mutation that hardlinks all files (to check that it keeps/fills the checksums)

**Comment by tavplubix** [comment]
No, it depends on what files have to be modified by a mutation. For example, when you run `ALTER ... UPDATE col1 = ...`, it rewrites `col1` and hardlinks other columns (and indexes that do not depend on `col1`). And `ALTER ... DROP COLUMN col1` hardlinks everything except `col1`.

**Pull request merged**